### PR TITLE
Refactor extracting preview image URLs from assets

### PR DESCRIPTION
**Pull request by Robinnnnn**
### Summary of Changes

https://www.loom.com/share/d6bd566927604159a509a21245fc1b2b?sid=79741c6d-07f6-4a03-bbe5-7b982c73aa0e

- This PR is a _prerequisite_ for streamlining how we handle `SyncingMedia` tokens across the app – specifically how it triggers false positives for broken-looking tokens.
- Screenshots are available for every file changed

**Core problems**
1. `SyncingMedia`, even if it has a fallback URL, will show `Loading...` in the editor sidebar and begin polling until it's processed. Usually a fallback is available to display much sooner.
2. Everywhere else in the app, `SyncingMedia` simply appears as invalid / broken. There are dozens of places where this occurs, such as the post composer, gallery page, feed, mobile composer, mobile gallery page, etc.
3. This has caused a host of downstream issues, including the fact that tokens look inconsistently broken across different surface areas.
4. One of our most widely reported user complaints is the fact that tokens look broken immediately upon syncing, when odds are that they're not. Worst case, the mobile app may cache the broken look across app refreshes!

image

**Why a solution is challenging**
1. We want to duplicate the polling + `Loading...` logic in most areas where NFTs are in a syncing state. We also want to display a fallback if available.
2. We display NFTs in a _lot_ of different places, each with different handlers
3. If we simply refactor all instances of `getVideoOrImageUrlForNftPreview` to better handle SyncingMedia and poll until it's resolved, it may result in duplicate / parallel polling for tokens that are rendered multiple times in the app
4. We need to set up a global context to de-duplicate polling
5. Which means we'll need to refactor `getVideoOrImageUrlForNftPreview` into a hook to access context
6. Which means we'll need to refactor upstream callers of that function to be a hook

### The Solution
1. This PR: Refactor all instances of `getVideoOrImageUrlForNftPreview` into a hook. The app should look and feel the same.
2. Next PR: Polling Context for syncing media. Ensure SyncingMedia fallback is displayed.
3. Do the same thing for mobile

### This PR

- The widely used `getVideoOrImageUrlForNftPreview` was refactored into `getPreviewImageUrlsInlineDangerously`. it should rarely be used directly, but performs the same job as its predecessor. in addition, it comes with `SyncingMedia` fallback support, returns an `isSyncing` state if no fallback is available, or an error if all else fails
- `useGetPreviewImagesWithPolling` wraps the above method with automatic error handling and polling for `SyncingMedia`. this removes the need for the caller to awkwardly pass in a `reportError` callback!
- `useGetSinglePreviewImage`: further wraps the above method, purely for convenience, given that in most cases users know what size image they want, and know what they would do without it.

image

**Further improvements**
- the `video` return type was deprecated from `getVideoOrImageUrlForNftPreview`, which makes the codebase a lot less horrible

image

image

**Sub-branches**
- `getOptimisticUserInfo` was converted to a hook in #1886
- web post composer-related files were refactored in #1888 

### Demo or Before/After Pics

- Screenshots included along each component file to ensure nothing broke

### Edge Cases

List any common edge cases that you have considered and tested.

### Testing Steps

Provide steps on how the reviewer can test the changes.

### Checklist

Please make sure to review and check all of the following:

- [ ] I've tested the changes and all tests pass.
- [ ] (if web) I've tested the changes on various desktop screen sizes to ensure responsiveness.
- [ ] (if mobile) I've tested the changes on both light and dark modes.

### CODE REVIEW

File: apps/web/pages/opengraph/collection/[collectionId].tsx [hunk 1]
```diff
@@ -4,7 +4,7 @@ import { graphql, useLazyLoadQuery } from 'react-relay';
 import { OpenGraphPreview } from '~/components/opengraph/OpenGraphPreview';
```
Comment (Robinnnnn) [inline apps/web/pages/opengraph/collection/[collectionId].tsx:4]: image

File: apps/web/pages/opengraph/gallery/[galleryId].tsx [hunk 2]
```diff
@@ -4,7 +4,7 @@ import { graphql, useLazyLoadQuery } from 'react-relay';
 import { OpenGraphPreview } from '~/components/opengraph/OpenGraphPreview';
```
Comment (Robinnnnn) [inline apps/web/pages/opengraph/gallery/[galleryId].tsx:4]: image

File: apps/web/pages/opengraph/nft/[tokenId].tsx [hunk 3]
```diff
@@ -4,7 +4,7 @@ import { graphql, useLazyLoadQuery } from 'react-relay';
 import { SingleOpenGraphPreview } from '~/components/opengraph/SingleOpenGraphPreview';
```
Comment (Robinnnnn) [inline apps/web/pages/opengraph/nft/[tokenId].tsx:4]: image

File: apps/web/pages/opengraph/post/[postId].tsx [hunk 4]
```diff
@@ -5,7 +5,7 @@ import { graphql, useLazyLoadQuery } from 'react-relay';
 import { PostOpenGraphPreview } from '~/components/opengraph/PostOpenGraphPreview';
```
Comment (Robinnnnn) [inline apps/web/pages/opengraph/post/[postId].tsx:5]: image

File: apps/web/pages/opengraph/user/[username].tsx [hunk 5]
```diff
@@ -4,7 +4,7 @@ import { graphql, useLazyLoadQuery } from 'react-relay';
 import { OpenGraphPreview } from '~/components/opengraph/OpenGraphPreview';
```
Comment (Robinnnnn) [inline apps/web/pages/opengraph/user/[username].tsx:4]: image

File: packages/shared/src/relay/getPreviewImageUrlsInlineDangerously.ts [hunk 6]
```diff
@@ -0,0 +1,258 @@
+import { graphql, readInlineData } from 'relay-runtime';
```
Comment (Robinnnnn) [inline packages/shared/src/relay/getPreviewImageUrlsInlineDangerously.ts:1]: a newer and better version of our existing `getVideoOrImageUrlForNftPreview`
1. handles `SyncingMedia` types
2. no longer returns the incredibly confusing `video` type, which addresses one of our oldest TODOs and alleviates a lot of legacy headaches

image

File: packages/shared/src/relay/useGetPreviewImageUrlsWithPolling.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
+import { useEffect } from 'react';
```
Comment (Robinnnnn) [inline packages/shared/src/relay/useGetPreviewImageUrlsWithPolling.ts:1]: the hook that wraps above method. comes with built-in polling and error reporting

File: apps/web/src/components/CommunityHolderGrid/CommunityHolderGridItem.tsx [hunk 8]
```diff
@@ -10,9 +10,8 @@ import { useModalActions } from '~/contexts/modal/ModalContext';
 import { CommunityHolderGridItemFragment$key } from '~/generated/CommunityHolderGridItemFragment.graphql';
```
Comment (Robinnnnn) [inline apps/web/src/components/CommunityHolderGrid/CommunityHolderGridItem.tsx:10]: image

File: apps/web/src/components/GalleryEditor/CollectionEditor/StagedNftImage.tsx [hunk 9]
```diff
@@ -5,14 +5,12 @@ import NftPreviewLabel from '~/components/NftPreview/NftPreviewLabel';
 import { StagedNftImageFragment$key } from '~/generated/StagedNftImageFragment.graphql';
```
Comment (Robinnnnn) [inline apps/web/src/components/GalleryEditor/CollectionEditor/StagedNftImage.tsx:5]: image

File: apps/web/src/components/GalleryEditor/CollectionEditor/StagedNftImageDragging.tsx [hunk 10]
```diff
@@ -7,10 +7,9 @@ import { StagedNftImageDraggingWithoutFallbackFragment$key } from '~/generated/S
 import { useImageFailureCheck } from '~/hooks/useImageFailureCheck';
```
Comment (Robinnnnn) [inline apps/web/src/components/GalleryEditor/CollectionEditor/StagedNftImageDragging.tsx:7]: image

File: apps/web/src/contexts/globalLayout/GlobalAnnouncementPopover/components/FeaturedCollectorCard.tsx [hunk 11]
```diff
@@ -11,7 +11,7 @@ import FollowButton from '~/components/Follow/FollowButton';
 import { FeaturedCollectorCardCollectionFragment$key } from '~/generated/FeaturedCollectorCardCollectionFragment.graphql';
```
Comment (Robinnnnn) [inline apps/web/src/contexts/globalLayout/GlobalAnnouncementPopover/components/FeaturedCollectorCard.tsx:11]: no screenshot for this one as it requires reinstating the GlobalAnnouncementPopover

will be easy enough to check if i messed anything up next time we work on that component

File: apps/web/src/components/ProfilePicture/ProfilePicture.tsx [hunk 12]
```diff
@@ -5,8 +5,9 @@ import { graphql, useFragment } from 'react-relay';
 import styled from 'styled-components';
```
Comment (Robinnnnn) [inline apps/web/src/components/ProfilePicture/ProfilePicture.tsx:5]: image

File: apps/web/src/components/Notifications/notifications/SomeoneCommentedOnYourPost.tsx [hunk 13]
```diff
@@ -1,4 +1,3 @@
-import { useMemo } from 'react';
```
Comment (Robinnnnn) [inline apps/web/src/components/Notifications/notifications/SomeoneCommentedOnYourPost.tsx:1]: image

File: apps/web/src/components/Notifications/notifications/SomeoneAdmiredYourPost.tsx [hunk 14]
```diff
@@ -1,4 +1,3 @@
-import { useMemo } from 'react';
```
Comment (Robinnnnn) [inline apps/web/src/components/Notifications/notifications/SomeoneAdmiredYourPost.tsx:1]: image

File: apps/web/src/components/Notifications/notifications/SomeoneAdmiredYourPost.tsx [hunk 15]
```diff
@@ -40,21 +39,25 @@ export default function SomeoneAdmiredYourPost({ notificationRef, onClose }: Pro
...
-    if (!token) return null;
-    return getVideoOrImageUrlForNftPreview({ tokenRef: token });
-  }, [token]);
 
   if (!firstAdmirer) {
-    reportError(`SomeoneAdmiredYourPostNotification id:${notification.dbid} was missing admirer`);
-    return null;
+    throw new ErrorWithSentryMetadata('SomeoneAdmiredYourPostNotification was missing admirer', {
+      id: notification.dbid,
+    });
```
Comment (Robinnnnn) [inline apps/web/src/components/Notifications/notifications/SomeoneAdmiredYourPost.tsx:50]: note: if a component should error, we should throw an error rather than call `reportError` directly in the render function; some parent component should catch the error in a boundary

in this case, a `ReportingErrorBoundary` was wrapped around the `Notification` component

File: apps/web/src/components/Notifications/Notification.tsx [hunk 16]
```diff
@@ -249,21 +250,23 @@ export function Notification({ notificationRef, queryRef, toggleSubView }: Notif
...
-          </HStack>
-          {!notification.seen && (
-            <UnseenDotContainer>
-              <UnseenDot />
-            </UnseenDotContainer>
-          )}
-        </StyledDotAndTimeAgo>
-      </HStack>
-    </Container>
+    <ReportingErrorBoundary fallback={null}>
```
Comment (Robinnnnn) [inline apps/web/src/components/Notifications/Notification.tsx:253]: see: https://github.com/gallery-so/gallery/pull/1885/files#r1316773290

File: apps/web/src/components/Notifications/notifications/NewTokens.tsx [hunk 17]
```diff
@@ -1,4 +1,4 @@
-import { useCallback, useMemo } from 'react';
```
Comment (Robinnnnn) [inline apps/web/src/components/Notifications/notifications/NewTokens.tsx:1]: image

File: packages/shared/src/relay/useGetPreviewImages.ts [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+    tokenRef
+  );
+
+  const result = getPreviewImageUrlsInlineDangerously({ tokenRef: token, preferStillFrameFromGif });
+
+  const reportError = useReportError();
+
+  if (result.type === 'error') {
+    reportError(result.error);
+  }
```
Comment (Robinnnnn) [inline packages/shared/src/relay/useGetPreviewImages.ts:77]: do this in a useEffect

File: apps/web/src/components/Feed/Socialize/CommentsModal/FeedEventsCommentsModal.tsx [hunk 19]
```diff
@@ -4,7 +4,7 @@ import { graphql, useFragment, usePaginationFragment } from 'react-relay';
 import { FeedEventsCommentsModalFragment$key } from '~/generated/FeedEventsCommentsModalFragment.graphql';
```
Comment (Robinnnnn) [inline apps/web/src/components/Feed/Socialize/CommentsModal/FeedEventsCommentsModal.tsx:4]: all components and files related to `getOptimisticUserInfo` refactoring was handled in a separate PR: #1886

File: apps/web/src/utils/useOptimisticUserInfo.ts [hunk 20]
```diff
@@ -0,0 +1,72 @@
+import { useMemo } from 'react';
```
Comment (Robinnnnn) [inline apps/web/src/utils/useOptimisticUserInfo.ts:1]: all components and files related to `getOptimisticUserInfo` refactoring was handled in a separate PR, which contains all the screenshots: #1886

**Review by kaitoo1 — APPROVED** [review]
WOW 🧠 🧠 🧠
